src/Storages/ExternalStream/Kafka/Producer.cpp

```diff
@@ -0,0 +1,44 @@
+#include <Common/logger_useful.h>
+#include <Storages/ExternalStream/Kafka/Producer.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int INVALID_CONFIG_PARAMETER;
+extern const int INVALID_SETTING_VALUE;
+}
+
+namespace RdKafka
+{
+
+/// Producer will take the ownership of `rk_conf`.
+Producer::Producer(rd_kafka_conf_t * rk_conf, UInt64 poll_timeout_ms, Poco::Logger * logger_) : logger(logger_)
+{
+    char errstr[512];
+    rk.reset(rd_kafka_new(RD_KAFKA_PRODUCER, rk_conf, errstr, sizeof(errstr)));
+    if (!rk)
+    {
+        /// librdkafka only take the ownership of `rk_conf` if `rd_kafka_new` succeeds,
+        /// we need to free it otherwise.
+        rd_kafka_conf_destroy(rk_conf);
```

Comment:
we should create and destroy the `rk_conf`  in same scope ? likes that
```c++
auto * new_rk_conf = rd_kafka_conf_dup(rk_conf);
rk.reset(rd_kafka_new(RD_KAFKA_PRODUCER, new_rk_conf, errstr, sizeof(errstr)));
if (!rk)
{
    rd_kafka_conf_destroy(new_rk_conf);
    ...
}
```

Reply:
Another exception-safe code is
```c++
std::unique_ptr<...> new_rk_conf{rd_kafka_conf_dup(rk_conf), rd_kafka_conf_destroy};
rk.reset(rd_kafka_new(RD_KAFKA_PRODUCER, new_rk_conf.get(), errstr, sizeof(errstr)));
if (rk)
    new_rk_conf.release(); /// the rk will take the ownership of `new_rk_config`
```

Reply:
I don't use unique_ptr here because it does not really bring much benefit in this case, if `rd_kafka_conf_dup` fails, it still fails.